cassandra/src/main/scala/akka/stream/contrib/cassandra/CassandraSource.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib.cassandra
+
+import akka.NotUsed
+import akka.stream._
+import akka.stream.scaladsl.Source
+import akka.stream.stage.{ OutHandler, GraphStageLogic, GraphStage }
+import com.datastax.driver.core.{ ResultSet, Session, Statement, Row }
+import com.google.common.util.concurrent.{ ListenableFuture, FutureCallback, Futures }
+import scala.collection.mutable
+import scala.concurrent.{ Future, Promise }
+import scala.collection.JavaConverters._
+import scala.util.{ Try, Failure, Success }
+
+class CassandraSource(stmt: Statement, session: Session) extends GraphStage[SourceShape[Row]] {
+  val out: Outlet[Row] = Outlet("CassandraSource")
+  override val shape: SourceShape[Row] = SourceShape(out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) {
+      val queue = mutable.Queue.empty[Row]
+      var maybeRs = Option.empty[ResultSet]
+
+      override def preStart(): Unit = {
+        implicit val ec = materializer.executionContext
+
+        val futRs = guavaFutToScalaFut(session.executeAsync(stmt))
+        val futRsCallback = getAsyncCallback[Try[ResultSet]] {
+          case Success(rs) =>
+            maybeRs = Some(rs)
+            queue.enqueue(rs.iterator().asScala.take(rs.getAvailableWithoutFetching).toSeq: _*)
```

Comment:
good suggestion, but I don't think there is a need for a buffer (queue) at all, see https://github.com/lagom/lagom/blob/master/persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cassandra/CassandraSessionImpl.scala#L272